### PR TITLE
Roll forward CC upgrade with fixes (#18552)

### PR DESCRIPTION
**Pull request by choumx**
See #18552.

### New changes:
- Remove `node_modules/` path from `js_module_root` option. It caused silent bundling failure with imports leaving undefined Closure module path vars e.g. `module$path$to$foo`.
- Document method for removing missing type paths in `swg.js`.
- Fix `react-dates` issue [1] by injecting it post-compile instead of a JS `import` in `amp-date-picker.js` and removed obsolete "erwinmHack" export hack. :)
- Fix a couple new type errors in `amp-date-picker.js`.

[1] CC v20171112 appears to DCE some of browserify's module loading boilerplate in `third_party/react-dates/bundle.js`. Specifically:

```js
"moment": [function(require, module, exports) {
    //! moment.js
    (function(global, factory) {
        // This outer ternary is DCE'd by CC v20171112, but not in v20170409.
        typeof exports === 'object' && typeof module !== 'undefined' 
            ? module.exports = factory() 
            : typeof define === 'function' && define.amd 
                ? define(factory) 
                : global.moment = factory()
    }(this, (function() {
        'use strict'; ...
```

### CODE REVIEW

File: build-system/get-dep-graph.js [hunk 1]
```diff
@@ -82,11 +82,16 @@ exports.getFlags = function(config) {
...
+    module_resolution: 'NODE',
     externs: config.externs,
     define: config.define,
     // Turn off warning for "Unknown @define" since we use define to pass
     // args such as FORTESTING to our runner.
     jscomp_off: ['unknownDefines'],
+    // checkVars: Demote "variable foo is undeclared" errors.
+    // moduleLoad: Demote "module not found" errors to ignore missing files
+    //     in type declarations in the swg.js bundle.
+    jscomp_warning: ['checkVars', 'moduleLoad'],
```
Comment (erwinmombay) [inline build-system/get-dep-graph.js]: did we need this for both compile and check-type passes?

Comment (choumx) [reply build-system/get-dep-graph.js]: Shouldn't we be consistent between both?

File: ads/adventive.js [hunk 2]
```diff
@@ -139,10 +139,13 @@ function getUrl(context, data, ns) {
...
-      isRecipeStale = !isRecipeLoaded ? true :
-        (Date.now() - ns.args[placementId].requestTime) > (60 * cacheTime),
-      needsRequest = !isRecipeLoaded || isRecipeStale;
+  const isRecipeLoaded = hasOwn(ns.args, 'placementId');
+  let isRecipeStale = true;
+  if (isRecipeLoaded) {
+    const info = ns.args[placementId];
+    isRecipeStale = (Date.now() - info['requestTime']) > (60 * cacheTime);
+  }
+  const needsRequest = !isRecipeLoaded || isRecipeStale;
```
Comment (kristoferbaxter) [inline ads/adventive.js]: This is clearer to follow, but the other methods have similar patterns as the previous code.

Was this change required to get the newer version of CC to correctly compile?

File: build-system/get-dep-graph.js [hunk 3]
```diff
@@ -206,7 +211,8 @@ exports.getBundleFlags = function(g) {
       throw new Error('Expect to build more than one bundle.');
     }
   });
-  flagsArray.push('--js_module_root', `${g.tmp}/node_modules/`);
+  // Do _not_ include 'node_modules/' in js_module_root with 'NODE' resolution
```
Comment (kristoferbaxter) [inline build-system/get-dep-graph.js]: Does the order matter here? Can these static items just be placed into the array when its created?

Perhaps a better followup change.

File: build-system/tasks/compile.js [hunk 4]
```diff
@@ -252,7 +251,7 @@ function compile(entryModuleFilenames, outputDir,
       'third_party/webcomponentsjs/ShadowCSS.js',
       'third_party/rrule/rrule.js',
       'third_party/react-dates/bundle.js',
-      'node_modules/dompurify/dist/purify.cjs.js',
+      'node_modules/dompurify/dist/purify.es.js',
```
Comment (kristoferbaxter) [inline build-system/tasks/compile.js]: Is this change part of moving to the new version or just convenience since there are already changes here.

File: build-system/tasks/dep-check.js [hunk 5]
```diff
@@ -175,8 +175,13 @@ function getGraph(entryModule) {
 
   // TODO(erwinm): Try and work this in with `gulp build` so that
   // we're not running browserify twice on travis.
-  const bundler = browserify(entryModule, {debug: true, deps: true})
-      .transform(babelify, {compact: false});
+  const bundler = browserify(entryModule, {debug: true})
+      .transform(babelify, {
+        compact: false,
+        // Transform files in node_modules since deps use ES6 export.
```
Comment (kristoferbaxter) [inline build-system/tasks/dep-check.js]: This is a significant change (especially before Babel 7), is this required?

File: ads/adventive.js (hunk 2 above)
Comment (choumx) [reply ads/adventive.js]: Yea, the real change was using brackets instead of dot for dereferencing `requestTime`. CC was complaining that the property did not exist.

File: build-system/get-dep-graph.js (hunk 3 above)
Comment (choumx) [reply build-system/get-dep-graph.js]: Not sure, Malte wrote this. :)

File: build-system/tasks/compile.js (hunk 4 above)
Comment (choumx) [reply build-system/tasks/compile.js]: I was having some trouble getting both Babel and CC to work when importing the CJS version. The ES version didn't have this issue but I'm not sure why.

File: build-system/tasks/dep-check.js (hunk 5 above)
Comment (choumx) [reply build-system/tasks/dep-check.js]: I believe so. I think being able to use ES module versions of dependencies is great consistency improvement, and may help `worker-dom` integration too. :)

**Comment by choumx** [comment]
Thanks for the reviews!

**Comment by choumx** [comment]
Looks like the upgrade increased bundle size. :(

[81.55KB](https://travis-ci.org/ampproject/amphtml/jobs/438813801) -> [82.1KB](https://travis-ci.org/ampproject/amphtml/jobs/438822190)

The original PR might not have exhibited this due to the `node_modules/` bundling issue.

File: ads/adventive.js (hunk 2 above)
Comment (kristoferbaxter) [reply ads/adventive.js]: Might be worth filing an issue to move the rest of this file over to the same approach.

File: build-system/get-dep-graph.js (hunk 3 above)
Comment (kristoferbaxter) [reply build-system/get-dep-graph.js]: Not a big deal. We can verify later.

File: build-system/tasks/compile.js (hunk 4 above)
Comment (kristoferbaxter) [reply build-system/tasks/compile.js]: Makes sense, thanks.

File: build-system/get-dep-graph.js (hunk 1 above)
Comment (erwinmombay) [reply build-system/get-dep-graph.js]: @choumx ideally. i just remembered the scenario you ran into originally with multi-pass, where the compilation worked but check-types failed. was that more an inconsistency with another option?

File: gulpfile.js [hunk 6]
```diff
@@ -1124,6 +1124,19 @@ function appendToCompiledFile(srcFilename, destFilePath) {
...
+    const file = fs.readFileSync(destFilePath, 'utf8');
+    const firstLineBreak = file.indexOf('\n');
+    const wrapperOpen = file.substr(0, firstLineBreak + 1);
+    const reactDates = fs.readFileSync(
+        'third_party/react-dates/bundle.js', 'utf8');
+    // Inject the bundle inside the standard AMP wrapper (after the first line).
+    const newSource = [
+      wrapperOpen, reactDates, file.substr(firstLineBreak + 1),
+    ].join('\n');
+    fs.writeFileSync(destFilePath, newSource, 'utf8');
```
Comment (choumx) [inline gulpfile.js]: @erwinmombay @cvializ FYI this is how I'm solving it for multi-pass.

**Comment by choumx** [comment]
Tested the normal push build steps locally (full compile and test etc.) and everything looks good after some final fixes. Merging! 🤞

File: gulpfile.js (hunk 6 above)
Comment (erwinmombay) [reply gulpfile.js]: @choumx did this actually work for single pass? i dont think single pass reaches this code path from the testing ive been doing on master. just double checking

Comment (choumx) [reply gulpfile.js]: I haven't tested it. This technique is certainly part of the solution for single-pass. We still need to either (1) add externs or (2) change all invocations into react, moment, etc. in AMP code to either use bracket notation.
